pkg/client/keystone/keystone.go

```diff
@@ -2,20 +2,94 @@ package keystone
 
 import (
 	"bytes"
+	"crypto/tls"
+	"crypto/x509"
 	"encoding/json"
 	"fmt"
+	"io"
 	"io/ioutil"
 	"net/http"
 
+	contrail "github.com/Juniper/contrail-operator/pkg/apis/contrail/v1alpha1"
+	"github.com/Juniper/contrail-operator/pkg/certificates"
 	"github.com/Juniper/contrail-operator/pkg/client/kubeproxy"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func NewClient(client *kubeproxy.Client) *Client {
-	return &Client{client: client}
+// NewClient prepares keystone client based on properties of passed keystone CRD.
+func NewClient(kubClient client.Client, scheme *runtime.Scheme, config *rest.Config, k *contrail.Keystone) (*Client, error) {
+	if k.Spec.ServiceConfiguration.ExternalAddress != "" {
+		caCertificate := certificates.NewCACertificate(kubClient, scheme, k, k.GetName())
+		caBundle, _ := caCertificate.GetCaCert()
+		return &Client{
+			Client:       newExtKeystoneClient(k.Spec.ServiceConfiguration.AuthProtocol, k.Spec.ServiceConfiguration.ExternalAddress, k.Spec.ServiceConfiguration.ListenPort, caBundle),
+			KeystoneConf: &k.Spec.ServiceConfiguration,
+		}, nil
+	}
+	proxy, err := kubeproxy.New(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create kubeproxy: %v", err)
+	}
+	if k.Spec.ServiceConfiguration.AuthProtocol == "https" {
+		return &Client{
+			Client:       proxy.NewSecureClientForService(k.Namespace, k.Name+"-service", k.Status.Port),
+			KeystoneConf: &k.Spec.ServiceConfiguration,
+		}, nil
+	}
+	return &Client{
+		Client:       proxy.NewClientForService(k.Namespace, k.Name+"-service", k.Status.Port),
+		KeystoneConf: &k.Spec.ServiceConfiguration,
+	}, nil
+}
+
+type keystoneClient interface {
+	NewRequest(method, path string, body io.Reader) (*http.Request, error)
+	Do(req *http.Request) (*http.Response, error)
 }
 
 type Client struct {
```

Comment:
this is confusing that the field in Client struct has the same name like the struct, I see that it was done like this before but could you please change it to sth like Proxy?

Reply:
Moreover why these fields are public?

Reply:
Renamed to Connector if that is ok? Wanted avoid confusion with kubeproxy client.

I made fields public because there is one case in command tests https://github.com/Juniper/contrail-operator/blob/73368d027fe6c60edd2a58772d066943a05c7f8e/test/e2e/aio/command_test.go#L270-L279 where connection to keystone through command pod is verifed. To handle such situation in newClient method of keyston client I would need to define additional parameters. I did not saw usage of it outside of test (rest of the requests goes through keystone service), but wanted to keep that test defined.

Reply:
that's completely fine